### PR TITLE
feat(subscription): add subscription budgets.

### LINKED ISSUE #64: feat: budget support
- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

## Description

Add capability to deploy budgets with subscriptions

### PR DESCRIPTION
**Pull request by nyanhp**
# Overview/summary

This PR adds the ability to define cost management budgets for a subscription, excluding filters.

## This PR fixes/adds/changes/removes

1. fixes #64

### Breaking changes

None

## Testing evidence

I am not familiar enough with Go to have added tests. I tested instead in my environment.

```
# module.lz_vending["subname"].module.subscription[0].azapi_resource.subscription_budget["budget1"] will be created
  + resource "azapi_resource" "subscription_budget" {
      + body                      = jsonencode(
            {
              + properties = {
                  + amount        = 150
                  + category      = "Cost"
                  + notifications = {
                      + eightypercent = {
                          + contactEmails = [
                              + "john@contoso.com",
                            ]
                          + contactGroups = []
                          + contactRoles  = []
                          + enabled       = true
                          + locale        = "en-us"
                          + operator      = "GreaterThan"
                          + threshold     = 80
                          + thresholdType = "Actual"
                        }
                      + forecast      = {
                          + contactEmails = [
                              + "john@contoso.com",
                            ]
                          + contactGroups = []
                          + contactRoles  = []
                          + enabled       = true
                          + locale        = "en-us"
                          + operator      = "GreaterThanOrEqualTo"
                          + threshold     = 290
                          + thresholdType = "Forecasted"
                        }
                      + ninetypercent = {
                          + contactEmails = [
                              + "john@contoso.com",
                            ]
                          + contactGroups = []
                          + contactRoles  = []
                          + enabled       = true
                          + locale        = "en-us"
                          + operator      = "GreaterThanOrEqualTo"
                          + threshold     = 90
                          + thresholdType = "Actual"
                        }
                    }
                  + timeGrain     = "Monthly"
                  + timePeriod    = {
                      + endDate   = "2027-12-31T23:59:59Z"
                      + startDate = "2024-05-01T00:00:00Z"
                    }
                }
            }
        )
      + id                        = (known after apply)
      + ignore_casing             = false
      + ignore_missing_property   = true
      + location                  = (known after apply)
      + name                      = "budget1"
      + output                    = (known after apply)
      + parent_id                 = "/subscriptions/nope"
      + removing_special_chars    = false
      + schema_validation_enabled = true
      + tags                      = (known after apply)
      + type                      = "Microsoft.Consumption/budgets@2021-10-01"
    }

  # module.lz_vending["subname"].module.subscription[0].azapi_resource.subscription_budget["budget2"] will be created
  + resource "azapi_resource" "subscription_budget" {
      + body                      = jsonencode(
            {
              + properties = {
                  + amount        = 1500
                  + category      = "Cost"
                  + notifications = {
                      + eightypercent = {
                          + contactEmails = [
                              + "john@contoso.com",
                            ]
                          + contactGroups = []
                          + contactRoles  = []
                          + enabled       = true
                          + locale        = "en-us"
                          + operator      = "GreaterThan"
                          + threshold     = 80
                          + thresholdType = "Actual"
                        }
                      + forecast      = {
                          + contactEmails = [
                              + "john@contoso.com",
                            ]
                          + contactGroups = []
                          + contactRoles  = []
                          + enabled       = true
                          + locale        = "en-us"
                          + operator      = "GreaterThanOrEqualTo"
                          + threshold     = 290
                          + thresholdType = "Forecasted"
                        }
                      + ninetypercent = {
                          + contactEmails = [
                              + "john@contoso.com",
                            ]
                          + contactGroups = []
                          + contactRoles  = []
                          + enabled       = true
                          + locale        = "en-us"
                          + operator      = "GreaterThanOrEqualTo"
                          + threshold     = 90
                          + thresholdType = "Actual"
                        }
                    }
                  + timeGrain     = "Annually"
                  + timePeriod    = {
                      + endDate   = "2027-12-31T23:59:59Z"
                      + startDate = "2024-05-01T00:00:00Z"
                    }
                }
            }
        )
      + id                        = (known after apply)
      + ignore_casing             = false
      + ignore_missing_property   = true
      + location                  = (known after apply)
      + name                      = "budget2"
      + output                    = (known after apply)
      + parent_id                 = "/subscriptions/nope"
      + removing_special_chars    = false
      + schema_validation_enabled = true
      + tags                      = (known after apply)
      + type                      = "Microsoft.Consumption/budgets@2021-10-01"
    }
```

Apply was successfully completed, e.g.:
image

## As part of this pull request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/terraform-azurerm-lz-vending/pulls)
- [x] Associated it with relevant [issues](https://github.com/Azure/terraform-azurerm-lz-vending/issues), for tracking and closure.
- [x] Run and `make fmt` & `make docs` to format your code and update documentation.
- [ ] Created unit and deployment tests and provided evidence.
- [x] Updated relevant and associated documentation.

### CODE REVIEW

File: modules/subscription/variables.tf [hunk 1]
```diff
@@ -196,3 +196,81 @@ variable "wait_for_subscription_before_subscription_operations" {
...
+  type = map(object({
+    amount            = number
+    time_grain        = optional(string, "Monthly")
+    time_period_start = string
+    time_period_end   = string
+    notifications = map(object({
+      enabled       = bool
+      operator      = string # EqualTo, GreaterThan, GreaterThanOrEqualTo
+      threshold     = number # 0-1000 percent
+      thresholdType = string # Actual, Forecasted
```
Comment (kewalaka) [inline modules/subscription/variables.tf:210]: Question for the team - does it make sense to have this interface "look" like azurerm - e.g. the input name would be camel-cased, like "threshold_type", and the ones below like "contact_emails" etc.

My thoughts are if we do switch out the implementation for an AzureRM version, the interface will remain the same.

Comment (jaredfholgate) [reply modules/subscription/variables.tf:210]: Agreed, the inputs for all terraform modules should be snake case.

File: modules/subscription/main.tf [hunk 2]
```diff
@@ -122,3 +122,26 @@ resource "azapi_resource_action" "subscription_cancel" {
...
+  for_each = var.subscription_budgets
+
+  type      = "Microsoft.Consumption/budgets@2021-10-01"
+  name      = each.key
+  parent_id = "/subscriptions/${local.subscription_id}"
+  body = jsonencode({
+    properties = {
+      amount        = each.value.amount
+      category      = "Cost"
+      notifications = each.value.notifications
```
Comment (jaredfholgate) [inline modules/subscription/main.tf:135]: As noted in the previous comments, this object should be expanded in order to use snake case variables.

File: variables.subscription.tf [hunk 3]
```diff
@@ -197,3 +197,55 @@ variable "wait_for_subscription_before_subscription_operations" {
...
+  type = map(object({
+    amount            = number
+    time_grain        = optional(string, "Monthly")
+    time_period_start = string
+    time_period_end   = string
+    notifications = map(object({
+      enabled       = bool
+      operator      = string # EqualTo, GreaterThan, GreaterThanOrEqualTo
+      threshold     = number # 0-1000 percent
+      thresholdType = string # Actual, Forecasted
```
Comment (jaredfholgate) [inline variables.subscription.tf:211]: As noted in other comments these could and should be snake case.

**Review by jaredfholgate — CHANGES_REQUESTED** [review]
I appreciate this is extra work, but the consistency of naming conventions will be beneficial in the longer term. Thanks

File: variables.subscription.tf (hunk 3 above)
Comment (nyanhp) [reply variables.subscription.tf]: added local to transform the notification map.

File: modules/subscription/main.tf (hunk 2 above)
Comment (nyanhp) [reply modules/subscription/main.tf:135]: added local to transform the notification map.

File: modules/subscription/variables.tf (hunk 1 above)
Comment (nyanhp) [reply modules/subscription/variables.tf]: added local to transform the notification map.

File: variables.subscription.tf [hunk 4]
```diff
@@ -197,3 +197,55 @@ variable "wait_for_subscription_before_subscription_operations" {
...
+  budget1 = {
+    amount            = 150
+    time_grain        = "Monthly"
+    time_period_start = "2024-01-01T00:00:00Z"
+    time_period_end   = "2027-12-31T23:59:59Z"
+    notifications = {
+      eightypercent = {
+        enabled       = true
+        operator      = "GreaterThan"
+        threshold     = "80"
```
Comment (jaredfholgate) [inline variables.subscription.tf:235]: Can we use a number rather than a string in the example?

File: variables.subscription.tf [hunk 5]
```diff
@@ -197,3 +197,55 @@ variable "wait_for_subscription_before_subscription_operations" {
...
+      threshold_type = string # Actual, Forecasted
+      contact_emails = optional(list(string), [])
+      contact_roles  = optional(list(string), [])
+      contact_groups = optional(list(string), [])
+      locale         = optional(string, "en-us")
+    }))
+  }))
+  default     = {}
+  description = <<DESCRIPTION
+The budgets to create for the subscription using the AzApi provider.
```
Comment (jaredfholgate) [inline variables.subscription.tf:220]: Is it possible to expand on these docs to provide field level explanations? Ideally it should be that case that someone consuming this module does not need to lookup the expected value formats elsewhere. You can link off to other docs for allowed values if there is a long list.

File: modules/subscription/variables.tf [hunk 6]
```diff
@@ -196,3 +196,81 @@ variable "wait_for_subscription_before_subscription_operations" {
...
+      operator       = string # EqualTo, GreaterThan, GreaterThanOrEqualTo
+      threshold      = number # 0-1000 percent
+      threshold_type = string # Actual, Forecasted
+      contact_emails = optional(list(string), [])
+      contact_roles  = optional(list(string), [])
+      contact_groups = optional(list(string), [])
+      locale         = optional(string, "en-us")
+    }))
+  }))
+  validation {
```
Comment (jaredfholgate) [inline modules/subscription/variables.tf:217]: Is it possible to extend the validation to include the notifications? Specifically `operator`, `threshold` and `threshold_type`?

**Review by jaredfholgate — COMMENTED** [review]
The code looks really good, just think we can improve the user experience a bit.

File: modules/budget/variables.tf [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  type        = number
+}
+
+variable "budget_time_grain" {
+  description = "The time grain of the budget."
+  type        = string
+  validation {
+    condition = contains(["Annually", "BillingAnnual", "BillingMonth", "BillingQuarter", "Monthly", "Quarterly"], var.budget_time_grain)
+
+    error_message = "Time period must be one of Annually, BillingAnnual, BillingMonth, BillingQuarter, Monthly, or Quarterly. BillingMonth, BillingQuarter, and BillingAnnual are only supported by WD customers."
```
Comment (matt-FFFFFF) [inline modules/budget/variables.tf:22]: What is a WD Customer?

**Review by matt-FFFFFF — COMMENTED** [review]
LGTM - one comment!

**Review by matt-FFFFFF — APPROVED** [review]

**Pull request merged**